src/main/java/com/greghaskins/spectrum/Suite.java

```diff
@@ -139,19 +124,44 @@ private void addChild(final Child child) {
   }
 
   void beforeAll(final Block block) {
-    this.beforeAll.addBlock(new IdempotentBlock(block));
+    addHook(new HookContext(before(new IdempotentBlock(block)), false, false, true));
   }
 
   void afterAll(final Block block) {
-    this.afterAll.addBlock(block);
+    addHook(new HookContext(after(block), false, true, false));
   }
 
   void beforeEach(final Block block) {
-    this.beforeEach.addBlock(block);
+    addHook(new HookContext(before(block), true, false, false));
   }
 
   void afterEach(final Block block) {
-    this.afterEach.addBlock(block);
+    addHook(new HookContext(after(block), true, false, false));
+  }
+
+  /**
+   * Adds a hook to be the first one executed before the block.
+   * This is the default. Hooks should be executed in the order they
+   * are declared in the test.
+   * @param hook to add
+   */
+  void addHook(final HookContext hook) {
+    this.hooks.addFirst(hook);
+  }
+
+  Hooks getHooksFor(final Child child) {
+    Hooks allHooks = parent.getInheritableHooks().plus(hooks);
+
+    return child instanceof Atomic ? allHooks.forAtomic() : allHooks.forNonAtomic();
```

Comment:
Tiny nit: is there a way to do this without `instanceof`? Seems like a possible design smell. Maybe `Child` just has an `isAtomic()` method. `Spec` and `CompositeTest` would say `true`, while `Suite` would say `false`.